### ISSUE TITLE
ADDITIONAL_DEPS environment variable has stopped working with ASP.NET Core 2.1

### ISSUE THREAD

**Issue by tobyash86** [aspnet/AspNetCore#3523]
### Description

My project is injecting an external library as a dependency for the CoreCLR profiler. It uses ADDITIONAL_DEPS environment variable to achieve that. Everything works well as long as ASP.NET Core 2.0 web application is profiled, but if I use exactly the same workflow with ASP.NET Core 2.1 my library is not found by the application.

### Diagnostics

This is what I have already tried:

- profiling .NET Core 2.0 application - OK
- profiling ASP.NET Core 2.0 application - OK
- profiling .NET Core 2.1 application - OK
- profiling ASP.NET Core 2.1 application using ADDITIONAL_DEPS env var - FAILURE
- profiling ASP.NET Core 2.1 application using --additional-deps switch - FAILURE

In each of these cases, I was using the same deps.json file and dependent library.

I have found out using procmon.exe that in case of ASP.NET Core 2.1 dotnet.exe behaves differently than in other cases. **Dependent assembly is NOT looked up in a location provided in \*.deps.json file, but only in the profiled application directory.**

Output from dottest.exe is as follows:

```
An assembly specified in the application dependencies manifest (hooks.deps.json) was not found:
    package: 'hooks', version: '1.0.0.0'
    path: 'C:/Program Files/Profiler/bin/hooks.dll'
```
I have double-checked assembly exists in this location.

And this is my \*.deps.json file:

```
{
"runtimeTarget": {
    "name": "hooks"
  },
  "targets": {
    "hooks": {
      "hooks/1.0.0.0": {
        "runtime": {
          "C:\\Program Files\\Profiler\\bin\\hooks.dll": {}
        }
      }
    }
  },
  "libraries": {
    "hooks/1.0.0.0": {
      "type": "reference",
      "serviceable": false,
      "sha512": ""
    }
  }
}
```

### Questions
**1. How can I use additional-deps with ASP.NET Core 2.1 properly?**
**2. Is there some other way to inject external assembly to ASP.NET Core application?**
**3. Is there at least a workaround for this issue?**

**Comment by Eilon** [comment]
@pranavkm / @pakrym - any thoughts on this?

**Comment by pakrym** [comment]
What the full path of `deps.json` file?
What is ADDITIONAL_DEPS set to?

**Comment by tobyash86** [comment]
Environment variable:
`
set DOTNET_ADDITIONAL_DEPS=c:\Users\MyUser\Documents\MyCompany\MyProduct\Profiler\hooks.deps.json
`
Physical path:
`
C:\Users\MyUser\Documents\MyCompany\MyProduct\Profiler\hooks.deps.json
`

My web application is hosted from a folder:
`
c:\inetpub\wwwroot\HW
`
and the only lookup for hooks assembly (based on procmon.exe output) is:
`
c:\inetpub\wwwroot\HW\hooks.dll
`

I will try to create a minimal repro for this issue.

**Comment by tobyash86** [comment]
Ok, I have managed to create a repro.

First, here is my dotnet --info output:
```
C:\Program Files (x86)\Microsoft Visual Studio 14.0>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.401
 Commit:    91b1c13032

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.1.7601
 OS Platform: Windows
 RID:         win7-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.401\

Host (useful for support):
  Version: 2.1.3
  Commit:  124038c13e

.NET Core SDKs installed:
  2.1.401 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

### Instructions for using repro
[ASP.NET_Core_Repro.zip](https://github.com/aspnet/Home/files/2370605/ASP.NET_Core_Repro.zip)

1. Zip contains compiled binaries, so first please trust the zip file by:
--- right click, select Properties from context menu
--- check Unblock option and click OK
2. Extract contents of zip file into C:\Temp
3. In C:\Temp\ASP.NET_Core_Repro there will be two batch files
--- running ASP_20_hooks should be successful
--- running ASP_21_hooks will fail with a message that hooks assembly cannot be found

**Comment by tobyash86** [comment]
### Update
While I was trying to prepare a repro it turned out that I was able to use ADDITIONAL_DEPS environment variable successfully when running:
`dotnet.exe [AppSrcDir]\bin\Debug\netcoreapp2.1\MyApp.dll`
but after running Publish from Visual Studio and running:
`dotnet.exe [AppDeployDir]\MyApp.dll`
ADDITIONAL_DEPS stopped working as expected.

Can you tell me what is the reason behind that?
Is it a bug or is it expected?
How to inject my assembly into deployed application then?

**Comment by tobyash86** [comment]
Is there any chance I will get some help from ASP.NET Core team regarding this issue?

**Comment by pakrym** [comment]
Is the app you are publishing portable or standalone?

**Comment by tobyash86** [comment]
![image](https://user-images.githubusercontent.com/34890669/45699780-18658000-bb6b-11e8-932e-8a904f098ef2.png)


I am not sure if this is the desired information. If it is not, then please tell me how can I check this.

**Comment by tobyash86** [comment]
### Another update
I was struggling to prepare a repro because the similar thing happened also while I was trying to prepare an example for ASP.NET Core 2.0. I could not use ADDITIONAL_DEPS with my ASP.NET Core 2.0 application because there was the same issue as in case of ASP.NET Core 2.1.

It turns out that target version of ASP.NET Core for a web application is not related to this error, it is the **installed runtime** that causes this issue.

### Repro

I have prepared a very simple [repro](https://github.com/aspnet/Home/files/2401027/hello-world-no-startup.2.0.zip) based on some example found on GitHub. The zipped example contains also debug binaries. It was built against ASP.NET Core 2.0.

Also, the second [package](https://github.com/aspnet/Home/files/2401057/hooks.zip) contains my sample assembly and its *.deps.json file to be used as a value for ADDITIONAL_DEPS environment variable.


**How to reproduce:**

**In each case, ADDITIONAL_DEPS should point to hooks.deps.json.**

Case 1: Having installed runtime and hosting for ASP.NET Core 2.0 (no 2.1 components installed). The attached example will work perfectly fine with dotnet.exe running hello-world-no-startup.dll assembly directly from the build folder. Also, after publishing it to IIS Server and running it from deployment folder it will work as expected without any errors.

Case 2: After installing runtime and hosting for ASP.NET Core 2.1 example will work properly with dotnet.exe running hello-world-no-startup.dll assembly directly from the build folder. After deploying the application to IIS Server, there will be an error preventing the application from starting, saying that hooks.dll defined in hooks.deps.json cannot be found.

Please let me know if you will need any help from my side.

**Comment by tobyash86** [comment]
Is there any progress regarding this issue. Can I provide any additional data to help to solve it?

**Comment by swaroop-sridhar** [comment]
Hi @tobyash86, I couldn't repro the failure on asp.net core 2.1. I was able to run the test on ASP.net2.1 on build and publish directories. I also tried publishing to IIS using webdeploy. Can you please paste the exact commands you tried, and the error?  
There's no reference to hooks.dll from the program. I didn't see an error even after setting the ADDITIONAL_DEPS in the environment.
Do you have the probing paths configured correctly?

**Comment by tobyash86** [comment]
Hi @swaroop-sridhar,
I will try to revalidate my repro on several different machines to eliminate an impact from my environment.
Also, I will do the checks which you have suggested and let you know.

> There's no reference to hooks.dll from the program. I didn't see an error even after setting the ADDITIONAL_DEPS in the environment.

No reference to hooks.dll is expected, as my scenario requires to inject that dll into any website dynamically.

Thanks for your engagement!

**Comment by vitek-karas** [comment]
You can also try to set COREHOST_TRACE=1 environment variable alongside your ADDITIONAL_DEPS. The process should then write detailed information about probing into the stderr which I assume ASP.NET somehow captures.

**Comment by tobyash86** [comment]
Sorry for 2 weeks delay, but I was busy with other topics.

Just a quick update: I have run the scenario on other machine and it worked properly. For now, the only difference I see is that machine 1 is Windows 7 machine, and the working one is Windows 10 machine.

For some reason on the Windows 10 machine, ADDITIONAL_DEPS environment variable is resolved properly and my hooks assembly is loaded by the CLR profiler. On Windows 7 machine lookup is done only in the website folder (after copying my assembly to that folder everything works as expected).

@swaroop-sridhar wrote:

> Do you have the probing paths configured correctly?

I doubt if this has anything to do with probing paths as the json file passed in ADDITIONAL_DEPS environment variable contains a full path to hooks assembly.

Tomorrow I will perform a more detailed comparison between the two machines and let you know about my findings.

**Comment by pkruk2** [comment]
Maybe this is dotnet related, not asp.net core? I found following https://github.com/dotnet/core-setup/blob/dbb1c1e55eab169f4af370e9a61b85a982ff1105/Documentation/design-docs/additional-deps.md which references https://github.com/dotnet/core-setup/issues/3889 which says: 

> Starting with netcore 2.1 runtime the logic to find deps files has changed.

The issue seems to be focused on additional deps pointing to a folder, but maybe there was a change (intended or not) which also affects the file case?

**Comment by tobyash86** [comment]
Was the provided repro successful on your side?

**Comment by swaroop-sridhar** [comment]
Hi @tobyash86, I haven't tried the new repro instructions yet; I'm working on another issue, and will get to it shortly. Thanks.

**Comment by pkruk2** [comment]
> Maybe this is dotnet related, not asp.net core?

I have checked and the same problem happens with normal .net core console application which is published.

- When it's published with Deployment Mode "Framework Dependent" and Target Runtime "Portable", you can workaround it by putting hooks.dll in the Target Location
- When it's published with Deployment Mode "Self-contained" and Target Runtime "win-x86", the workaround does not work

**Comment by swaroop-sridhar** [comment]
Since this is a dotnet core issue, let's mode the discussion to the appropriate repo here: https://github.com/dotnet/core-setup/issues/4811 

Thanks.

**Comment by pkruk2** [comment]
I will try to prepare repro this week or early next week.

**Comment by tobyash86** [comment]
Closing.

**Comment by tobyash86** [comment]
I have found out that using DOTNET_ADDITIONAL_DEPS requires also specifying additionalprobingpath parameter. Using both of them works well.

There is another problem though... Can I specify additionalprobingpath as an environment varable? My goal is to set it dynamically for ASP.NET Core application.

**Comment by swaroop-sridhar** [comment]
@tobyash86 `additionalprobingpaths` can be set in the `runtimeconfig.json` file or the command line parameter. But there is no option to set it as an environment variable. 
`additionalprobingpaths` is processed at host startup time.

**Comment by tobyash86** [comment]
I think that DOTNET_ADDITIONAL_DEPS is working similarly, but can be specified as an environment variable. I will post a feature request for that.

**Comment by tobyash86** [comment]
Also, I think this issue cannot be closed, because --additional-deps behaves differently for .NET Core than for ASP.NET Core.

I have prepared a series of batch files running .NET Core processes with different settings. Please check my [repro](https://drive.google.com/open?id=1tTvpx6c0NYGXk9pzJicBXhEZ_1-SQ2ZD).

On my machine ASP_HelloWorldWithDeps_FAIL.bat behaves differently than CON_SampleAppWithDeps_OK - please note that both of them does not have --additionalprobingpath specified.

I have Runtime 2.2.4 and SDK 2.2.203 installed. I had to uninstall any other SDK/runtime versions to reproduce this issue. Please refer to README.txt in repro package for more details.

**Comment by anurse** [comment]
@tobyash86 ASP.NET Core has no impact on how `ADDITIONAL_DEPS` is handled, it's handled entirely by the dotnet host. https://github.com/dotnet/core-setup is a more appropriate repo for bugs/feature requests there so they are visible to the team that develops that component. https://github.com/dotnet/core-setup/issues/4811 appears to be covering some of this.

Closing as this is not in an ASP.NET Core component.
